Core/GDCore/Project/ObjectFolderOrObject.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * GDevelop Core
+ * Copyright 2008-2023 Florian Rival (Florian.Rival@gmail.com). All rights
+ * reserved. This project is released under the MIT License.
+ */
+#include "GDCore/Project/ObjectFolderOrObject.h"
+
+#include <memory>
+
+#include "GDCore/Project/Object.h"
+#include "GDCore/Project/ObjectsContainer.h"
+#include "GDCore/Serialization/SerializerElement.h"
+#include "GDCore/Tools/Log.h"
+
+using namespace std;
+
+namespace gd {
+
+ObjectFolderOrObject ObjectFolderOrObject::badObjectFolderOrObject;
+
+ObjectFolderOrObject::ObjectFolderOrObject() : folderName("__NULL") {}
+ObjectFolderOrObject::ObjectFolderOrObject(gd::String folderName_,
+                                           ObjectFolderOrObject* parent_)
+    : folderName(folderName_), parent(parent_) {}
+ObjectFolderOrObject::ObjectFolderOrObject(gd::Object* object_,
+                                           ObjectFolderOrObject* parent_)
+    : object(object_), parent(parent_) {}
+ObjectFolderOrObject::~ObjectFolderOrObject() {}
+
+bool ObjectFolderOrObject::HasObjectNamed(const gd::String& name) {
+  if (IsFolder()) {
+    return std::any_of(
+        children.begin(),
+        children.end(),
+        [&name](
+            std::unique_ptr<gd::ObjectFolderOrObject>& objectFolderOrObject) {
+          return objectFolderOrObject.get()->HasObjectNamed(name);
+        });
+  }
+  return object->GetName() == name;
+}
+ObjectFolderOrObject& ObjectFolderOrObject::GetObjectNamed(
+    const gd::String& name) {
+  if (!IsFolder() && object->GetName() == name) {
+    return *this;
+  }
+  if (IsFolder()) {
+    for (std::size_t j = 0; j < children.size(); j++) {
+      ObjectFolderOrObject& foundInChild = children[j]->GetObjectNamed(name);
+      if (&(foundInChild) != &badObjectFolderOrObject) {
+        return foundInChild;
+      }
+    }
+  }
+  return badObjectFolderOrObject;
+}
+
+void ObjectFolderOrObject::SetFolderName(const gd::String& name) {
+  if (!IsFolder()) return;
+  folderName = name;
+}
+
+ObjectFolderOrObject& ObjectFolderOrObject::GetChildAt(std::size_t index) {
+  return *children[index];
```

Comment:
I wondered if I should do that but I ended up mimicking what ObjectsContainer does with its objects. So I'm not sure what to do

Reply:
I would say: let's go for the safe option for this class and return badObjectFolderOrObject.
I prefer a logical error in the editor rather than a crash in most of these classes defining a project structure.